src/main/java/com/bandwidth/sqs/queue/buffer/KeyedTaskBuffer.java

```diff
@@ -20,64 +23,113 @@
+    }
+
+    /**
+     * Construct a KeyedTaskBuffer with all parameters
+     * @param scheduledExecutorService - The executor used for task execution when a batch's wait time has expired
+     * @param maxBufferSize - The maximum amount of task data to hold before flushing the batch and executing it
+     * @param maxWait - The maximum amount of time to hold a batch before flushing it regardless of size
+     * @param task - The task that executes the batched data
+     */
+    public KeyedTaskBuffer(ScheduledExecutorService scheduledExecutorService, int maxBufferSize,
+                           Duration maxWait, Task<K, D> task) {
+        this.scheduledExecutorService = scheduledExecutorService;
         this.maxBufferSize = maxBufferSize;
         this.maxWait = maxWait;
         this.task = task;
     }
 
-    public synchronized void addData(K key, D data) {
-        List<D> buffer = buffers.get(key);
-        if (buffer == null) {
-            TimerTask timerTask = new TimerExpiredTask(key);
-            buffer = new ArrayList<>();
-            buffers.put(key, buffer);
-            timer.schedule(timerTask, maxWait.toMillis());
-        }
-        buffer.add(data);
-        processBatchIfNeeded(key, false);
-    }
 
-    private synchronized void processBatchIfNeeded(K key, boolean expired) {
-        if (!buffers.containsKey(key)) {
-            return;
+    /**
+     * Add task data to a keyed buffer.  If the buffer is full the batch will be executed immediately
+     * on the calling thread. Otherwise the task will be executed by the scheduleExecutorService when
+     * the maxWait timeout has elapsed.
+     * @param key - Task Key.  Tasks with common keys are batched
+     * @param data - Task Data
+     */
+    public void addData(final K key, D data) {
+        List<D> readyBatch;
+
+        synchronized (this) {
+            List<D> buffer = buffers.get(key);
+            if (buffer == null) {
+                buffer = new ArrayList<>();
+                buffers.put(key, buffer);
+
+                scheduledExecutorService.schedule(() -> {
+                    try {
+                        processExpiredBatch(key);
+                    } catch (Exception e) {
+                        LOG.error("Exception running task with key {}", key,  e);
+                    }
+                }, maxWait.toMillis(), TimeUnit.MILLISECONDS);
+            }
+            buffer.add(data);
+
+            //check if the batch is ready while holding the lock to prevent batches from
+            //going above the maximum size
+            readyBatch = fetchAndRemoveBatchIfReady(key, false);
         }
-        List<D> buffer = buffers.get(key);
-        boolean isMaxSize = buffer.size() >= maxBufferSize;
-        if (isMaxSize || expired) {
-            buffers.remove(key);
-            task.run(key, buffer);
+
+        //Execute the ready batch without holding the lock.
+        if (readyBatch != null) {
+            task.run(key, readyBatch);
```

Comment:
why are we doing it on the calling thread?

Reply:
1) it already worked this way so it's safer to not break this 2) It prevents a context switch and having to schedule on a separate thread when under load 3) It allows the thread pool to be smaller because we can more efficiently use Jetty's (the caller's) thread pool).

Reply:
I wonder if there is a problem, though, with subjecting the calling thread [pool] to what could be such a drastic variation in how long those threads take to run
If most of Jetty's threads call this method and it returns immediately, but occasionally one of those threads is actually stuck in this method until a Redis call completes, what are the consequences of that?

Reply:
That functionality already exists as I noted in point #1.